### DISCUSSION TITLE
Can you add a feature to export/convert PyCaret model to ONNX format?  This would enable deploying models created with PyCaret to any platform.

### DISCUSSION THREAD

**Discussion by tsereno** [pycaret/pycaret#1269] [Ideas]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. 

**Describe the solution you'd like**
A clear and concise description of what you want to happen, ideally taking into consideration the existing toolbox design, classes and methods. 

**Describe alternatives you've considered**

**Additional context**

**Comment by ngupta23** [comment]
@moezali1 wrote [an article](https://towardsdatascience.com/deploy-pycaret-models-on-edge-devices-with-onnx-runtime-c6d060a2e1a6) on it a few days back. Please refer to it to see how it can be done.

**Comment by pycaret** [comment]
Here is the link to the tutorial: https://towardsdatascience.com/deploy-pycaret-models-on-edge-devices-with-onnx-runtime-c6d060a2e1a6

**Comment by buhrmann** [comment]
Hi, I was wondering if you could elaborate a bit on your support for Onnx. If I understand it correctly, you rely on skl2onnx being able to convert your models automatically. Does this work for all the possible transformers in pycaret, or would what would be necessary to make a more complicated pycaret pipeline Onnx-compatible?

**Comment by harrisonogle** [comment]
@buhrmann 

> Does this work for all the possible transformers in pycaret

Unfortunately, no. We can check the `models` table to see which models will work with the approach in the article, which recommends `skl2onnx` to convert sklearn models produced by PyCaret.

```python
from pycaret.internal.tabular import models
model_df: pd.DataFrame = models().copy()
model_df['skl2onnx_compatible'] = model_df.Reference.str.startswith('sklearn.')
print(model_df.skl2onnx_compatible.reset_index())
```

You can see which models produced by PyCaret are compatible with `skl2onnx`:

```
          ID  skl2onnx_compatible
0         lr                 True
1      lasso                 True
2      ridge                 True
3         en                 True
4        lar                 True
5       llar                 True
6        omp                 True
7         br                 True
8        ard                 True
9        par                 True
10    ransac                 True
11        tr                 True
12     huber                 True
13        kr                 True
14       svm                 True
15       knn                 True
16        dt                 True
17        rf                 True
18        et                 True
19       ada                 True
20       gbr                 True
21       mlp                 True
22  lightgbm                False
23     dummy                 True
```

So currently, it looks like all models are compatible with `skl2onnx` besides LightGBM (the reason I ended up here! :D). Maybe the article was written at a time when all models were strictly sklearn.

If you want to take the easy way out, you could specify `exclude=['lightgbm']` when training your models (in `compare_models`, for example), but then you would not be leveraging LightGBM.

From `skl2onnx` [official documentation main page](https://onnx.ai/sklearn-onnx/), we find the recommendation for LightGBM (no anchor links, but it's under **Related converters** heading):

> sklearn-onnx only converts models from scikit-learn. [onnxmltools](https://github.com/onnx/onnxmltools) can be used to convert models for libsvm, lightgbm, xgboost. Other converters can be found on [github/onnx](https://github.com/onnx/), [torch.onnx](https://pytorch.org/docs/stable/onnx.html), [ONNX-MXNet API](https://mxnet.incubator.apache.org/api/python/contrib/onnx.html), [Microsoft.ML.Onnx](https://www.nuget.org/packages/Microsoft.ML.Onnx/)…

Interestingly enough, the [onnxmltools docs](https://github.com/onnx/onnxmltools) actually start off by revealing that they support conversion of scikit-learn models using a wrapper of `skl2onnx` converter... this means we can actually just use `onnxmltools` if we want.

> what would be necessary to make a more complicated pycaret pipeline Onnx-compatible?

The `onnxmltools` provides two methods we can use to convert any model (currently) produced by PyCaret. Those are:
- [convert_sklearn](https://github.com/onnx/onnxmltools/blob/158f2a2d340eb0c1dd2c6d29a9e365f3ad1d6794/docs/examples/plot_convert_sklearn.py), and
- [convert_lightgbm](https://github.com/onnx/onnxmltools/blob/158f2a2d340eb0c1dd2c6d29a9e365f3ad1d6794/docs/examples/plot_convert_lightgbm.py)

Unfortunately, I was unable to find a function which resolves the model type and uses the correct converter, so that means it's up to us to implement that logic. I created a full example earlier out of necessity which I can share:

```python
from pycaret.datasets import get_data
from pycaret.regression import setup, compare_models, pull, get_config
from onnxmltools.convert import convert_sklearn, convert_lightgbm
from skl2onnx.algebra.type_helper import guess_initial_types
from skl2onnx import to_onnx

# Train a model on example data according to the official PyCaret "Getting Started" example.
data = get_data('insurance')
setup(data, target='charges', silent=True, verbose=False)
best = compare_models(verbose=False)

# Get the ID of the model type (lr, knn, lightgbm, etc.) from the scoring table output of `compare_models` 
best_model_type_id = pull().iloc[0].name

print(best)
print(best_model_type_id)

# Get a sample of our training set.
X_sample = get_config('X_train')[:1].to_numpy()
# We can't hand `convert_lightgbm` a training sample, like we do in `skl2onnx.to_onnx`, so we must
# use `guess_initial_types` to infer types (this is used internally by `skl2onnx.to_onnx`).
initial_types = guess_initial_types(X_sample, None)
# Use the correct function from `onnxmltools` to convert the model.
onnx_model = (convert_lightgbm(best, initial_types=initial_types)
              if best_model_type_id == 'lightgbm'
              else to_onnx(best, X_sample))  # we could use `onnxmltools.convert.convert_sklearn` here instead,
                                             # but since we imported `skl2onnx` anyways, let's use it.
print(onnx_model)
```

Of course this solution is less than ideal! It requires us to maintain more code, which goes against the "low-code" appeal of PyCaret. It would be nice to have some advice on best practices for ONNX conversion of models produced by PyCaret. @ngupta23, would you be able to provide some guidance?

**Comment by explainable-ai** [comment]
When converting the model to onnx, does it convert the whole pipeline or just the model?

**Reply by kopczynski-9livesdata** [reply]
@explainable-ai I'm afraid that the article https://towardsdatascience.com/deploy-pycaret-models-on-edge-devices-with-onnx-runtime-c6d060a2e1a6 describes converting only the model, without the pipeline.
My attempt to convert the model with the pipeline ended up with errors like 
```
MissingShapeCalculator( skl2onnx.common.exceptions.MissingShapeCalculator: Unable to find a shape calculator for type '<class 'pycaret.internal.preprocess.transformers.TransformerWrapper'>'. It usually means the pipeline being converted contains a transformer or a predictor with no corresponding converter implemented in sklearn-onnx. If the converted is implemented in another library, you need to register the converted so that it can be used by sklearn-onnx (function update_registered_converter). If the model is not yet covered by sklearn-onnx, you may raise an issue to https://github.com/onnx/sklearn-onnx/issues to get the converter implemented or even contribute to the project. If the model is a custom model, a new converter must be implemented. Examples can be found in the gallery.
```
